File: CodeChef_Problems/Feb Long 21/TEAMNAME.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long

void printMap(map<string,vector<char>>mp)
{
    for(auto x : mp)
    {
        cout<<x.first<<" -> ";

        for(int i = 0; i < x.second.size(); i++)
            cout<<x.second[i]<<" ";
        
        cout<<endl;
    }
    cout<<endl;
}

int uniqueElements(vector<char>A, vector<char>B)
{
    set<char>S;

    for(int i = 0; i < A.size(); i++)
        S.insert(A[i]);
    
    for(int i = 0; i < B.size(); i++)
        S.insert(B[i]);

    return S.size();
}

void solve()
{
    int N;
    cin>>N;

    vector<string>V; //input strings

    for(int i = 0; i < N; i++)
    {
        string inp;
        cin>>inp;
        V.push_back(inp);
    }

    map<string,vector<char>>mp;

    /*

    for hell,bell,best,test the map would be suffix -> prefixes

    ell -> h , b;
    est -> b , t;

    */

   for(int i = 0; i < N; i++)
   {
       string str = V[i];

       string suffix = "";

        for(int j = 1; j < str.size(); j++)
            suffix += str[j];
        
        mp[suffix].push_back(str[0]);
   }

   //printMap(mp);

   int total = 0;

   for(auto it1 = mp.begin(); it1 != mp.end(); it1++)
   {

       for(auto it2 = mp.begin(); it2 != mp.end(); it2++)
       {
           int sz = uniqueElements(it1 -> second, it2 -> second);

            //cout<<sz<<endl;

           total += (sz - (it1 -> second).size())*(sz - (it2 -> second).size());
       }
   }

   cout<<total<<endl;
}

int main()
{
    int T;
    cin>>T;

    while(T--)
        solve();

    return 0;
}

File: CodeChef_Problems/Feb Long 21/TEAMNAME.py
def uniqueElements(A,B):
    return len(set(A + B))

def solve():
    N = int(input())
    S = input().split()

    sp_dict = {} #suffix -> prefix dict

    '''
    i/p : hell bell best test
    ell -> h , b
    est -> b , t
    '''

    for s in S:
        st = s[1 : ]

        if st in sp_dict:
            sp_dict[st].append(s[0])
        else:
            sp_dict[st] = [s[0]]

    suffix_keys = list(sp_dict.keys())

    total = 0

    for it1 in range(len(sp_dict)):
        for it2 in range(len(sp_dict)):
            sz = uniqueElements(sp_dict[suffix_keys[it1]], sp_dict[suffix_keys[it2]])
            total = total + (sz - len(sp_dict[suffix_keys[it1]]))*(sz - len(sp_dict[suffix_keys[it2]]))

    print(total)

def main():
    T = int(input())

    while(T):
        solve()
        T = T - 1

if __name__ == "__main__":
    main()
